### PR TITLE
Don't use "char[1] at end of struct" trick

### PR DESCRIPTION
**Pull request by richsalz**
An official interepretation of the C Standard says that this is not
strictly conforming. Consulting a high authority, Dennis Ritchie has
called this technique "unwarranted chumminess with the compiler."

Sadly, this increases memory fragmentation/malloc use.  Such is the
price of portability and correctness..

I was trying to track down the intermittent failures in the `threads_test`
With this change I haven't been able to get it to fail; before it would fail
about 20% of the time.

### CODE REVIEW

**Comment by richsalz** [comment]
Sigh.  When I do the same config as the no-cache build, it still fails :(  This is still a more-correct thing to do, however.  Still working on the failure.

**Comment by paulidale** [comment]
This doesn't seem like it would fix any threading problems.  It might hide them.
The structure in question is read only after creation, so should be immune to threading concerns.

The hash table isn't however.

**Comment by rsbeckerca** [comment]
@richsalz @paulidale I happen to know of an ornery system with finicky threads that might bring stuff to light. Would you like a test on it with PUT or SPT threads?

**Comment by levitte** [comment]
I don't understand how the "char[1] at end of struct" has an impact on threads.  Care to explain?

File: crypto/property/property_string.c [hunk 1]
```diff
@@ -117,11 +117,10 @@ static PROPERTY_STRING *new_property_string(const char *s,
     PROPERTY_STRING *ps = OPENSSL_malloc(sizeof(*ps) + l);
 
     if (ps != NULL) {
-        memcpy(ps->body, s, l + 1);
```
Comment (t8m) [inline crypto/property/property_string.c:120]: If the `s` does not need to be aligned, and it does not need to be IMO. You can avoid the extra allocation call even without having the char body[1] member in the struct.

Just:
```
void *mem = OPENSSL_malloc(sizeof(*ps) + l + 1);
PROPERTY_STRING *ps = mem;
char *body = mem;

body += sizeof(*ps);
memcpy(body, s, l + 1);
ps->s = body;
```

Comment (richsalz) [reply crypto/property/property_string.c:120]: I still don't like that trick.  If you want to close this PR and do that, go ahead.

**Comment by richsalz** [comment]
It doesn't have an impact on threads.  I found it while trying to figure out why the no-cache build fails the thread test every once in a while, and I went "that's poor code."

**Comment by richsalz** [comment]
> @richsalz @paulidale I happen to know of an ornery system with finicky threads that might bring stuff to light. Would you like a test on it with PUT or SPT threads?

Thanks for the offer, but I don't think that's needed.  The threads_test fails with a no-cache build on common platforms, too.

**Comment by richsalz** [comment]
> This doesn't seem like it would fix any threading problems. It might hide them.
> The structure in question is read only after creation, so should be immune to threading concerns.
> 
> The hash table isn't however.

I don't know how it would hide them.  Sorry for my initial comment leading everyone astray.  I agree this doesn't *fix* any threads problems.  I was mislead because I thought I had config'd a no-cache build and I was wrong.

So is the hash table the problem?  Not being cleaned up on exit?

**Comment by paulidale** [comment]
Insertion into the hash table is what I meant.  Once primed, it should be fine for concurrent read only access.

**Comment by richsalz** [comment]
This is a code-portability issue.  Any interest?

**Comment by rsbeckerca** [comment]
> 
> 
> This is a code-portability issue. Any interest?

I'm not seeing any alignment issue here.
```
struct t1 {
        char a[3];
        char b[];
};

struct t2 {
        char a[3];
        char b[1];
};
```
The address of `b` is aligned odd in both cases, while the size is 3 and 4 respectively. That's ANSI, I think. Is that what is expected?

**Comment by richsalz** [comment]
Your `t1` is part of C11, and OpenSSL targets c99 where the behavior is probably implementation-defined.  See for example https://stackoverflow.com/questions/36577094/array-of-size-0-at-the-end-of-struct

**Comment by rsbeckerca** [comment]
> Your `t1` is part of C11, and OpenSSL targets c99 where the behavior is probably implementation-defined.  See for example https://stackoverflow.com/questions/36577094/array-of-size-0-at-the-end-of-struct
> 

Actually, I was using c99. I don't even have c11 on that box.

**Comment by richsalz** [comment]
> Actually, I was using c99.

What can I tell you?  The "[]" is not part of C99, it is part of C11. It is arguably undefined behavior in C99.

**Comment by rsbeckerca** [comment]
> 
> 
> > Actually, I was using c99.
> 
> What can I tell you? The "[]" is not part of C99, it is part of C11. It is arguably undefined behavior in C99.

Weird. I get a warning in c89 and none in c99, but works in both cases - not `[0]`, but `[]` as the last field has worked in every compiler since I started using C in (mumble mumble 82). Granted that was in BSD 4.1 that may have had mods.

With that said, `dirent.h` has used a mis-defined `d_name` for epochs. So I'm wondering how big a red herring this is.

**Comment by richsalz** [comment]
>So I'm wondering how big a red herring this is.

What, nonstop or in general?

It seems to work on all platforms of interest so far.  It is not correct C code.  I have no opinion on nonstop.

**Comment by rsbeckerca** [comment]
> 
> 
> > So I'm wondering how big a red herring this is.
> 
> What, nonstop or in general?
> 
> It seems to work on all platforms of interest so far. It is not correct C code. I have no opinion on nonstop.

I guess it's only a red-herring on NonStop. Noticed that `d_name` is defined correctly in GNU even if it is allocated short. Perhaps that's the correct solution. On NonStop, it is defined as follows, even though names can be much larger than 8.
```
char      d_name[8];   /* indeterminate size */
```
Otherwise it is:
```
char d_name[NAME_MAX + 1];
```
But I think the struct allocation containing `d_name` is not necessarily that long. `gdb` thinks it is, but the actual memory content looks like there are more interesting parts beyond the terminating null. So perhaps declare the maximum size but only allocate the used size instead of declaring it short and allocating more. It certainly makes debuggers happier.

**Comment by levitte** [comment]
We target C99? That's news to me...

**Comment by rsbeckerca** [comment]
> 
> 
> We target C99? That's news to me...

c89, in 2019, Issue #8012; 2018, #7556; and I can't find the cases where someone stated c99 is standard for 3.0.0 but I'm pretty sure it was... might have been @levitte ;)

**Comment by richsalz** [comment]
No, C89 a persistent typo/off-by-one error on my part.  Whatever.  Still really not legal C and should be fixed.

**Comment by richsalz** [comment]
Should I even bother to rebase this, or is the project okay with the current state of the code?

**Comment by levitte** [comment]
I frankly don't see much of a point with this.  This trick is not our invention, and as far as I've experienced, it works everywhere.

Let me say it like this: don't go looking into the ASN.1 template code :wink:

**Comment by richsalz** [comment]
> I frankly don't see much of a point with this. This trick is not our invention, and as far as I've experienced, it works everywhere.

Then why do any warnings at all?  Just get rid of `gcc_devteam_warn` in Configure.

> Let me say it like this: don't go looking into the ASN.1 template code 😉

I have.  Look for the comments about `-Wextended-offsetof` in Configure.  Except for that, the code is clean.

Feel free to close this and I'll shut up.

**Pull request closed without merging**